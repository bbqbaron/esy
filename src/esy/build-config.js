/**
 * @flow
 */

import type {BuildSpec, BuildConfig} from './types';
import * as path from 'path';

export function createConfig(
  params: {
    storePath: string,
    sandboxPath: string,
  },
): BuildConfig {
  const {storePath, sandboxPath} = params;
  const localStorePath = path.join(
    sandboxPath,
    'node_modules',
    '.cache',
    '_esy',
    'store',
  );
  const genPath = (build: BuildSpec, tree: string, segments: string[]) => {
    if (build.shouldBePersisted) {
      return path.join(storePath, tree, build.id, ...segments);
    } else {
      return path.join(localStorePath, tree, build.id, ...segments);
    }
  };

  const buildConfig: BuildConfig = {
    sandboxPath,
    storePath,
    localStorePath,
    getSourcePath: (build: BuildSpec, ...segments) => {
      return path.join(buildConfig.sandboxPath, build.sourcePath, ...segments);
    },
    getRootPath: (build: BuildSpec, ...segments) => {
      if (build.mutatesSourcePath) {
        return genPath(build, '_build', segments);
      } else {
        return path.join(buildConfig.sandboxPath, build.sourcePath, ...segments);
      }
    },
    getBuildPath: (build: BuildSpec, ...segments) => genPath(build, '_build', segments),
    getInstallPath: (build: BuildSpec, ...segments) =>
      genPath(build, '_insttmp', segments),
    getFinalInstallPath: (build: BuildSpec, ...segments) =>
      genPath(build, '_install', segments),
  };
  return buildConfig;
}
