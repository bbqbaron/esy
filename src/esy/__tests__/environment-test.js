/**
 * @flow
 */

import type {Build} from '../build-repr';
import {
  expandWithScope,
  renderWithScope,
  calculate,
  printEnvironment,
} from '../environment';
import * as Config from '../build-config';

function build({name, exportedEnv, dependencies}): Build {
  return {
    id: name,
    name,
    version: '0.1.0',
    sourcePath: name,
    dependencies,
    exportedEnv,
    mutatesSourcePath: false,
    shouldBePersisted: true,
    command: null,
    errors: [],
  };
}

const ocaml = build({
  name: 'ocaml',
  exportedEnv: {
    CAML_LD_LIBRARY_PATH: {
      val: '$ocaml__lib/ocaml:$CAML_LD_LIBRARY_PATH',
      scope: 'global',
    },
  },
  dependencies: [],
});

const ocamlfind = build({
  name: 'ocamlfind',
  exportedEnv: {
    CAML_LD_LIBRARY_PATH: {
      val: '$ocamlfind__lib/ocaml:$CAML_LD_LIBRARY_PATH',
      scope: 'global',
    },
  },
  dependencies: [ocaml],
});

const lwt = build({
  name: 'lwt',
  exportedEnv: {
    CAML_LD_LIBRARY_PATH: {
      val: '$lwt__lib/ocaml:$CAML_LD_LIBRARY_PATH',
      scope: 'global',
    },
  },
  dependencies: [ocaml],
});

const config = Config.createConfig({
  sandboxPath: '<sandboxPath>',
  storePath: '<storePath>',
});

describe('renderWithScope()', function() {
  test('simple replacement: $var syntax', function() {
    const scope = new Map([['name', {value: 'World'}]]);
    expect(renderWithScope('Hello, $name!', scope)).toEqual({
      rendered: 'Hello, World!',
    });
  });

  test('multiple replacements', function() {
    const scope = new Map([['name', {value: 'World'}]]);
    expect(renderWithScope('Hello, $name + $name!', scope)).toEqual({
      rendered: 'Hello, World + World!',
    });
  });

  test('missing in scope', function() {
    const scope = new Map([]);
    expect(renderWithScope('Hello, $unknown!', scope)).toEqual({
      rendered: 'Hello, $unknown!',
    });
  });
});

describe('expandWithScope()', function() {
  test('simple replacement: $var syntax', function() {
    const scope = new Map([['name', {value: 'World'}]]);
    expect(expandWithScope('Hello, $name!', scope)).toEqual({
      rendered: 'Hello, World!',
    });
  });

  test('simple replacement: ${var} syntax', function() {
    const scope = new Map([['name', {value: 'World'}]]);
    expect(expandWithScope('Hello, ${name}!', scope)).toEqual({
      rendered: 'Hello, World!',
    });
  });

  test('multiple replacements', function() {
    const scope = new Map([['name', {value: 'World'}]]);
    expect(expandWithScope('Hello, $name + $name!', scope)).toEqual({
      rendered: 'Hello, World + World!',
    });
  });

  test('missing in scope', function() {
    const scope = new Map([]);
    expect(expandWithScope('Hello, $unknown!', scope)).toEqual({
      rendered: 'Hello, !',
    });
  });

  test('expansion with fallback to default value', function() {
    const scope = new Map([]);
    expect(expandWithScope('Hello, ${unknown:-Me}!', scope)).toEqual({
      rendered: 'Hello, Me!',
    });
  });
});

describe('calculate()', function() {
  expect.addSnapshotSerializer({
    test(val) {
      return val.id && val.name && val.dependencies && val.exportedEnv;
    },
    print(val) {
      return `Build { id: "${val.id}" }`;
    },
  });

  test('build with no exports', function() {
    const app = build({
      name: 'app',
      exportedEnv: {},
      dependencies: [],
    });
    expect(calculate(config, app)).toMatchSnapshot();
  });

  test('build with local exports', function() {
    const app = build({
      name: 'app',
      exportedEnv: {
        app__var: {val: 'hello'},
      },
      dependencies: [],
    });
    expect(calculate(config, app)).toMatchSnapshot();
  });

  test('build with global exports', function() {
    const app = build({
      name: 'app',
      exportedEnv: {
        APP: {val: 'hello', scope: 'global'},
      },
      dependencies: [],
    });
    expect(calculate(config, app)).toMatchSnapshot();
  });

  test('build with global export referencing built-in', function() {
    const app = build({
      name: 'app',
      exportedEnv: {
        APP: {val: 'hello, $app__name', scope: 'global'},
      },
      dependencies: [],
    });
    expect(calculate(config, app)).toMatchSnapshot();
  });

  test('build with global export referencing built-in (cur-version)', function() {
    const app = build({
      name: 'app',
      exportedEnv: {
        APP: {val: 'hello, $cur__name', scope: 'global'},
      },
      dependencies: [],
    });
    expect(calculate(config, app)).toMatchSnapshot();
  });

  test('build with (dep with local export)', function() {
    const dep = build({
      name: 'dep',
      exportedEnv: {
        dep__var: {val: 'hello'},
      },
      dependencies: [],
    });
    const app = build({
      name: 'app',
      exportedEnv: {},
      dependencies: [dep],
    });
    expect(calculate(config, app)).toMatchSnapshot();
  });

  test('build with (dep with local export) with global export referencing dep built-in', function() {
    const dep = build({
      name: 'dep',
      exportedEnv: {},
      dependencies: [],
    });
    const app = build({
      name: 'app',
      exportedEnv: {
        APP: {val: 'hello, $dep__name', scope: 'global'},
      },
      dependencies: [dep],
    });
    expect(calculate(config, app)).toMatchSnapshot();
  });

  test('build with (dep with local export) with global export referencing dep local export', function() {
    const dep = build({
      name: 'dep',
      exportedEnv: {
        dep__var: {val: 'hello'},
      },
      dependencies: [],
    });
    const app = build({
      name: 'app',
      exportedEnv: {
        APP: {val: '$dep__var, world', scope: 'global'},
      },
      dependencies: [dep],
    });
    expect(calculate(config, app)).toMatchSnapshot();
  });

  test('build with (dep with local export) with local export referencing dep built-in', function() {
    const dep = build({
      name: 'dep',
      exportedEnv: {},
      dependencies: [],
    });
    const app = build({
      name: 'app',
      exportedEnv: {
        app__var: {val: 'hello, $dep__name'},
      },
      dependencies: [dep],
    });
    expect(calculate(config, app)).toMatchSnapshot();
  });

  test('build with (dep with local export) with local export referencing dep local export', function() {
    const dep = build({
      name: 'dep',
      exportedEnv: {
        dep__var: {val: 'hello'},
      },
      dependencies: [],
    });
    const app = build({
      name: 'app',
      exportedEnv: {
        app_var: {val: '$dep__var, world'},
      },
      dependencies: [dep],
    });
    expect(calculate(config, app)).toMatchSnapshot();
  });

  test('build with (dep with (dep with local export) with local export)', function() {
    const depOfDep = build({
      name: 'dep-of-dep',
      exportedEnv: {
        dep_of_dep__var: {val: 'hello'},
      },
      dependencies: [],
    });
    const dep = build({
      name: 'dep',
      exportedEnv: {
        dep__var: {val: 'hello'},
      },
      dependencies: [depOfDep],
    });
    const app = build({
      name: 'app',
      exportedEnv: {},
      dependencies: [dep],
    });
    expect(calculate(config, app)).toMatchSnapshot();
  });

  test('concatenating global exports', function() {
    const app1 = build({
      name: 'app',
      exportedEnv: {},
      dependencies: [ocamlfind, ocaml],
    });
    expect(calculate(config, app1)).toMatchSnapshot();
    // check that order is deterministic (b/c of topo sort order of deps)
    const app2 = build({
      name: 'app',
      exportedEnv: {},
      dependencies: [ocaml, ocamlfind],
    });
    // TODO: uncomment this and make it pass
    //expect(calculate(config, app1)).toEqual(calculate(config, app2));
  });

  test('concatenating global exports (same level exports)', function() {
    const app = build({
      name: 'app',
      exportedEnv: {},
      dependencies: [ocamlfind, lwt],
    });
    expect(calculate(config, app)).toMatchSnapshot();
  });

  test('concatenating global exports (same level exports + package itself)', function() {
    const app = build({
      name: 'app',
      exportedEnv: {
        CAML_LD_LIBRARY_PATH: {
          val: '$app__lib:$CAML_LD_LIBRARY_PATH',
          scope: 'global',
        },
      },
      dependencies: [ocamlfind, lwt],
    });
    expect(calculate(config, app)).toMatchSnapshot();
  });
});

describe('printEnvironment()', function() {
  test('printing environment', function() {
    const app = build({
      name: 'app',
      exportedEnv: {
        CAML_LD_LIBRARY_PATH: {
          val: '$app__lib:$CAML_LD_LIBRARY_PATH',
          scope: 'global',
        },
      },
      dependencies: [ocamlfind, lwt],
    });
    const {env} = calculate(config, app);
    expect(printEnvironment(env)).toMatchSnapshot();
  });
});
