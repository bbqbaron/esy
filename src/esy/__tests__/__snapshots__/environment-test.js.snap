// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`printEnvironment() printing environment 1`] = `
"# Esy Sandbox
export OCAMLFIND_CONF=\\"<storePath>/_build/app/_esy/findlib.conf\\"
export PATH=\\"<storePath>/_install/ocaml/bin:<storePath>/_install/ocamlfind/bin:<storePath>/_install/lwt/bin:$PATH\\"
export MAN_PATH=\\"<storePath>/_install/ocaml/man:<storePath>/_install/ocamlfind/man:<storePath>/_install/lwt/man:$MAN_PATH\\"
# app@0.1.0 app
export cur__name=\\"app\\"
export cur__version=\\"0.1.0\\"
export cur__root=\\"<sandboxPath>/app\\"
export cur__depends=\\"ocamlfind lwt\\"
export cur__target_dir=\\"<storePath>/_build/app\\"
export cur__install=\\"<storePath>/_insttmp/app\\"
export cur__bin=\\"<storePath>/_insttmp/app/bin\\"
export cur__sbin=\\"<storePath>/_insttmp/app/sbin\\"
export cur__lib=\\"<storePath>/_insttmp/app/lib\\"
export cur__man=\\"<storePath>/_insttmp/app/man\\"
export cur__doc=\\"<storePath>/_insttmp/app/doc\\"
export cur__stublibs=\\"<storePath>/_insttmp/app/stublibs\\"
export cur__toplevel=\\"<storePath>/_insttmp/app/toplevel\\"
export cur__share=\\"<storePath>/_insttmp/app/share\\"
export cur__etc=\\"<storePath>/_insttmp/app/etc\\"
export CAML_LD_LIBRARY_PATH=\\"<storePath>/_install/app/lib:<storePath>/_install/lwt/lib/ocaml:<storePath>/_install/ocamlfind/lib/ocaml:<storePath>/_install/ocaml/lib/ocaml\\""
`;
